pyqtorch/utils.py

```diff
@@ -142,3 +143,31 @@ def density_mat(state: Tensor) -> Tensor:
     state = torch.permute(state, batch_first_perm).reshape(batch_size, 2**n_qubits)
     undo_perm = (1, 2, 0)
     return torch.permute(torch.einsum("bi,bj->bij", (state, state.conj())), undo_perm)
+
+
+def promote_ope(operator: Tensor, target: int, n_qubits: int) -> Tensor:
```

Comment:
Are you sure this is the best way to apply a given operator? Padding the smaller one with identities makes it quite expensive. I feel like there should be a smart way to apply the small operator on the correct subspace. But I am not sure, I have never really done any density matrix coding.

Reply:
Maybe this reference might have some cool ideas: https://arxiv.org/abs/2103.13962
And they have some code here: https://github.com/cmendl/density-matrix-bloch-qcircuit/tree/master

But tbh if there is not some easily implementable idea in that paper or somewhere else, I think your approach is also ok for an initial prototype. We can then optimize it later.

Reply:
In theoretical terms, that's how it's done, but indeed I hadn't considered the complexity. So, I'll think about a less expensive version.
However, before making any major modifications to a function, I'd like to finish the implementation of the Kraus operator (#148) completely, even if it means revisiting each of the functions later.
Also, if you want to discuss the` promote_ope()` function, please do so on this PR (#155). Here, it's just there for `apply_ope_ope()` to work and it's not the final version.